htlcswitch/alias_store.go

```diff
@@ -0,0 +1,190 @@
+	backend kvdb.Backend
+}
+
+func newAliasStore(db kvdb.Backend) *aliasStore {
+	return &aliasStore{backend: db}
+}
+
+// putPeerAlias stores the peer's alias SCID once we learn of it in the
+// funding_locked message.
+func (s *aliasStore) putPeerAlias(chanID lnwire.ChannelID,
+	alias lnwire.ShortChannelID) error {
+
+	return kvdb.Update(s.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(invoiceAliasBucket)
+		if err != nil {
+			return err
+		}
+
+		var scratch [8]byte
+		byteOrder.PutUint64(scratch[:], alias.ToUint64())
+		return bucket.Put(chanID[:], scratch[:])
+	}, func() {})
+}
+
+// getPeerAlias retrieves a peer's alias SCID by the channel's ChanID.
+func (s *aliasStore) getPeerAlias(chanID lnwire.ChannelID) (
+	lnwire.ShortChannelID, error) {
+
+	var alias lnwire.ShortChannelID
+
+	err := kvdb.Update(s.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(invoiceAliasBucket)
+		if err != nil {
+			return err
+		}
+
+		aliasBytes := bucket.Get(chanID[:])
+		if aliasBytes == nil {
+			return nil
+		}
+
+		alias = lnwire.NewShortChanIDFromInt(
+			byteOrder.Uint64(aliasBytes),
+		)
+		return nil
+	}, func() {})
+
+	return alias, err
+}
+
+// requestAlias returns a new ALIAS ShortChannelID to the caller by allocating
+// the next un-allocated ShortChannelID. The starting ShortChannelID is
+// 10000:0:0 and the ending ShortChannelID is 262143:16777215:65535. This gives
+// roughly 2^58 possible ALIAS ShortChannelIDs which ensures this space won't
+// get exhausted.
+func (s *aliasStore) requestAlias() (lnwire.ShortChannelID, error) {
+	var nextAlias lnwire.ShortChannelID
+
+	err := kvdb.Update(s.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(aliasAllocBucket)
+		if err != nil {
+			return err
+		}
+
+		lastBytes := bucket.Get(lastAliasKey)
+		if lastBytes == nil {
+			// If the key does not exist, then we can write the
+			// startingAlias to it.
+			nextAlias = startingAlias
+
+			var scratch [8]byte
+			byteOrder.PutUint64(scratch[:], nextAlias.ToUint64())
+			return bucket.Put(lastAliasKey, scratch[:])
+		}
+
+		// Otherwise the key does exist so we can convert the retrieved
+		// lastAlias to a ShortChannelID and use it to assign the next
+		// ShortChannelID. This next ShortChannelID will then be
+		// persisted in the database.
+		lastScid := lnwire.NewShortChanIDFromInt(
+			byteOrder.Uint64(lastBytes),
+		)
+		nextAlias = getNextScid(lastScid)
+
+		var scratch [8]byte
+		byteOrder.PutUint64(scratch[:], nextAlias.ToUint64())
+		return bucket.Put(lastAliasKey, scratch[:])
+	}, func() {})
+	if err != nil {
+		return nextAlias, err
+	}
+
+	return nextAlias, nil
+}
+
+// getNextScid is a utility function that returns the next SCID for a given
+// alias SCID. The BlockHeight ranges from [10000, 262143], the TxIndex ranges
+// from [1, 16777215], and the TxPosition ranges from [1, 65535].
+func getNextScid(last lnwire.ShortChannelID) lnwire.ShortChannelID {
```

Comment:
Interesting mapping here..seems a bit easier than trying to detect "overflow" and "carries" when doing addition. Will need to meditate on the current approach a bit.